src/microbe_stage/Microbe.Interior.cs

```diff
@@ -1323,16 +1332,58 @@ private void OnOrganelleRemoved(PlacedOrganelle organelle)
         cachedRotationSpeed = null;
         organelleMaxRenderPriorityDirty = true;
 
-        Compounds.Capacity = organellesCapacity;
+        UpdateCompoundBagCapacities();
     }
 
     /// <summary>
     ///   Recomputes storage from organelles, used after loading a save
     /// </summary>
     private void RecomputeOrganelleCapacity()
     {
-        organellesCapacity = organelles!.Sum(o => o.StorageCapacity);
-        Compounds.Capacity = organellesCapacity;
+        foreach (PlacedOrganelle organelle in organelles!)
+            UpdateCapacity(organelle, false);
+
+        UpdateCompoundBagCapacities();
+    }
+
+    /// <summary>
+    ///   Updates <see cref="organellesCapacity"/>
+    ///   and <see cref="additionalCompoundCapacities"/>
+    ///   to take into account the addition or removal of an organelle
+    /// </summary>
+    /// <param name="organelle">The organelle being placed or removed</param>
+    /// <param name="negative">
+    ///   Should be true if the organelle is being removed,
+    ///   otherwise false
+    /// </param>
+    private void UpdateCapacity(PlacedOrganelle organelle, bool negative)
+    {
+        int sign = negative ? -1 : 1;
+
+        if (organelle.Upgrades?.CustomUpgradeData is StorageComponentUpgrades storage &&
+            storage.SpecializationFor != null)
+        {
+            Compound specialization = storage.SpecializationFor;
+            additionalCompoundCapacities.TryGetValue(specialization, out var existing);
+            additionalCompoundCapacities[specialization] = existing + organelle.StorageCapacity * 2 * sign;
+        }
+        else
+        {
+            organellesCapacity += organelle.StorageCapacity * sign;
+        }
+    }
+
+    /// <summary>
+    ///   Updates <see cref="Compounds"/> to adjust for changes in
+    ///   <see cref="organellesCapacity"/> and
+    ///   <see cref="additionalCompoundCapacities"/>
+    /// </summary>
+    private void UpdateCompoundBagCapacities()
+    {
+        Compounds.NominalCapacity = organellesCapacity;
+
+        foreach (var entry in additionalCompoundCapacities)
+            Compounds.SetCapacityForCompound(entry.Key, entry.Value + organellesCapacity);
```

Comment:
This doesn't remove special capacities that are now gone from the cell (for example if the player switches a vacuole from glucose to ammonia, they get the extra capacity for both as the old value is never seemingly removed). There needs to be a method in the compound bag to reset the special capacities before they are re-applied to ensure no outdated values exist.

Reply:
Regarding this, the feature still works as intended. This is because `additionalCompoundCapacities` for in your example glucose is set to zero when the capacity is begin calculated. This means that while the entry exists in the dictionary, the value is zero. I have also play-tested this and the feature works as intended.

Reply:
Ah, you mean with `UpdateCapacity`? I guess that is true, though with the way I'm refactoring the microbe stage the "organelle removed" callback will be entirely removed, so could you futureproof your code by doing that my suggestion? That way I don't accidentally introduce this bug when trying to fix merge conflicts with my branch.